lib/charms/mlops_libs/v0/k8s_service_info.py

```diff
@@ -0,0 +1,358 @@
+from pydantic import BaseModel
+
+# The unique Charmhub library identifier, never change it
+LIBID = "f5c3f6cc023e40468d6f9a871e8afcd0"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+# Default relation and interface names. If changed, consistency must be kept
+# across the provider and requirer.
+DEFAULT_RELATION_NAME = "k8s-service-info"
+DEFAULT_INTERFACE_NAME = "k8s-service"
+REQUIRED_ATTRIBUTES = ["svc_name", "svc_port"]
+
+logger = logging.getLogger(__name__)
+
+
+class KubernetesServiceInfoRelationError(Exception):
+    """Base exception class for any relation error handled by this library."""
+
+    pass
+
+
+class KubernetesServiceInfoRelationMissingError(KubernetesServiceInfoRelationError):
+    """Exception to raise when the relation is missing on either end."""
+
+    def __init__(self):
+        self.message = "Missing relation with a k8s service info provider."
+        super().__init__(self.message)
+
+
+class KubernetesServiceInfoRelationDataMissingError(KubernetesServiceInfoRelationError):
+    """Exception to raise when there is missing data in the relation data bag."""
+
+    def __init__(self, message):
+        self.message = message
+        super().__init__(self.message)
+
+
+class KubernetesServiceInfoObject(BaseModel):
+    """Representation of a Kubernetes Service info object.
+
+    Args:
+        name: The name of the Service
+        port: The port of the Service
+    """
+
+    name: str
+    port: str
+
+
+class KubernetesServiceInfoRequirer(Object):
+    """Implement the Requirer end of the Kubernetes Service Info relation.
+
+    Observes the relation events and get data of a related application.
+
+    Args:
+        charm (CharmBase): the provider application
+        refresh_event: (list, optional): list of BoundEvents that this manager should handle.  Use this to update
+                       the data sent on this relation on demand.
+        relation_name (str, optional): the name of the relation
+
+    Attributes:
+        charm (CharmBase): variable for storing the requirer application
+        relation_name (str): variable for storing the name of the relation
+    """
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        refresh_event: Optional[Union[BoundEvent, List[BoundEvent]]] = None,
+        relation_name: Optional[str] = DEFAULT_RELATION_NAME,
+    ):
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+        self._requirer_wrapper = KubernetesServiceInfoRequirerWrapper(
+            self.charm, self.relation_name
+        )
+
+        self.framework.observe(self.charm.on.leader_elected, self._get_data)
+
+        self.framework.observe(self.charm.on[self.relation_name].relation_created, self._get_data)
```

Comment:
I think these should be `relation_changed` and `relation_broken`, like [here](https://github.com/canonical/resource-dispatcher/blob/6802b737d2d0053d2e6da4d4dbb0a2a55530d979/lib/charms/resource_dispatcher/v0/kubernetes_manifests.py#L181)

Reply:
I think that this wasn't caught by tests suggests we're missing a test.  I see the wrappers are unit tested, but not the `KubernetesServiceInfo*` classes afaict

Reply:
> I think that this wasn't caught by tests suggests we're missing a test. I see the wrappers are unit tested, but not the `KubernetesServiceInfo*` classes afaict

They are tested, but I think we are missing a test case for these events. Will update.

Reply:
> I think that this wasn't caught by tests suggests we're missing a test. I see the wrappers are unit tested, but not the `KubernetesServiceInfo*` classes afaict

They are tested, but I think what you are requesting is different, and therefore not tested.

It's not clear to me what is your request. Is it:
a) Instead of just observing `relation_created` we should observe `relation_changed` and `relation_broken` and bound the `get_data` method to those events?
b) Instead of just observing `relation_created` we should observe `relation_changed` and `relation_broken` and emit custom events?

If it is a), then I agree we could do that. If it's b) I think that'll change a bit the implementation details of this class, potentially affecting both the requirer/provider classes.

Reply:
Maybe we should discuss in person as it might be a bit easier to iterate.  I'm not 100% solid on these things either

At minimum, this should observe `relation_changed` and `relation_broken`.  `relation_created` only means there is an app on the other end, but says nothing about there being data on the relation, so it isn't needed here.  Also I don't think `leader_elected` is important here.  We block non-leaders from writing to application data (because ops prevents that), but reading should be fine.

imo, I like (b) more than (a) because it means the lib handles the low level `relation_*` events, and then emits higher level events that the charm needs to pay attention to (like how the data libs work).  That abstraction feels good to me - it hides the low-level behavior.  But in this case, where the low-level events map 1:1 with something the charm wants to hear about, I agree it isn't so important.  

I think there is a problem though with the current implementation.  On `relation_created`, the `KubernetesServiceInfoRequirer.get_data()` will fire and the databag will be empty.  This will result in `raise KubernetesServiceInfoRelationDataMissingError(...)` happening within ops (where we cannot catch that exception), putting the charm into error.  I think this would be caught with a test like:
```python
def test_requirer_get_data_if_empty(requirer_charm_harness):
    """Assert the relation data is as expected."""
    # Initial configuration
    requirer_charm_harness.set_model_name("test-model")
    requirer_charm_harness.set_leader(True)

    # Instantiate KubernetesServiceInfoRequirer class
    requirer_charm_harness.charm._k8s_svc_info_requirer = KubernetesServiceInfoRequirer(
        requirer_charm_harness.charm, relation_name=TEST_RELATION_NAME
    )

    requirer_charm_harness.begin()

    # Add (empty)
    rel_id = requirer_charm_harness.add_relation(TEST_RELATION_NAME, "app")

    # I think this will raise an error
```

I don't think a charm can recover from this.  Essentially the issue is a self-enclosed lib (eg: one that handles events itself) can never raise an Exception (because we can never catch them in the charm code).  But then its fine if you call the `KubernetesServiceInfoRequirerWrapper.get_data()` directly from your charm code, because you can catch the exceptions.

Reply:
afaict, `KubernetesServiceInfoRequirer` isn't a critical component of everything.  What about we just leave it out for now?  Charms can use the `KubernetesServiceInfoRequirerWrapper` and we could add this in later.  wdyt?

Reply:
I have updated this piece of code. I think that doing what you suggest makes sense and serves as an example of how our libraries should look like. Please let me know what you think.